sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala

```diff
@@ -179,8 +194,12 @@ class FileScanRDD(
             case c: ColumnarBatch => new ColumnarBatch(
               Array.tabulate(c.numCols())(c.column) ++ createMetadataColumnVector(c),
               c.numRows())
-            case u: UnsafeRow => projection.apply(new JoinedRow(u, metadataRow))
-            case i: InternalRow => new JoinedRow(i, metadataRow)
+            case u: UnsafeRow =>
+              rowIndexUpdater.foreach(_.update(u, metadataRow))
+              projection.apply(new JoinedRow(u, metadataRow))
```

Comment:
This will be slow slow slow. It is better to use `JoinedProjection` here.

Reply:
As opposed to `row.update()` or as opposed to `JoinedRow(...)`?